File: app/src/main/java/com/app/ralaunch/manager/ThemeManager.java
package com.app.ralaunch.manager;

import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import androidx.appcompat.app.AppCompatActivity;
import com.app.ralaunch.data.SettingsManager;
import com.app.ralaunch.utils.AppLogger;

import java.io.File;

/**
 * 主题管理器
 * 负责管理主题应用
 */
public class ThemeManager {
    private final AppCompatActivity activity;
    private final SettingsManager settingsManager;
    
    public ThemeManager(AppCompatActivity activity) {
        this.activity = activity;
        this.settingsManager = SettingsManager.getInstance(activity);
    }
    
    /**
     * 从设置中应用主题
     */
    public void applyThemeFromSettings() {
        int themeMode = settingsManager.getThemeMode(); // 0=跟随系统, 1=深色, 2=浅色
        
        switch (themeMode) {
            case 0: // 跟随系统
                androidx.appcompat.app.AppCompatDelegate.setDefaultNightMode(
                    androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
                break;
            case 1: // 深色模式
                androidx.appcompat.app.AppCompatDelegate.setDefaultNightMode(
                    androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_YES);
                break;
            case 2: // 浅色模式
                androidx.appcompat.app.AppCompatDelegate.setDefaultNightMode(
                    androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_NO);
                break;
        }
    }
    
    /**
     * 应用背景设置
     */
    public void applyBackgroundFromSettings() {
        String type = settingsManager.getBackgroundType();
        Drawable background = null;

        switch (type) {
            case "default":
                // 使用默认背景（从主题中获取）
                try {
                    int resId = activity.getResources().getIdentifier("bg_main", "drawable", activity.getPackageName());
                    if (resId != 0) {
                        background = activity.getResources().getDrawable(resId, activity.getTheme());
                    }
                } catch (Exception e) {
                    AppLogger.error("ThemeManager", "无法加载默认背景: " + e.getMessage());
                }
                // 如果加载失败，使用白色背景
                if (background == null) {
                    background = new ColorDrawable(0xFFFFFFFF);
                }
                break;
            case "color":
                // 使用纯色背景
                int color = settingsManager.getBackgroundColor();
                background = new ColorDrawable(color);
                break;
            case "image":
                // 使用图片背景
                String imagePath = settingsManager.getBackgroundImagePath();
                if (imagePath != null && !imagePath.isEmpty()) {
                    File imageFile = new File(imagePath);
                    if (imageFile.exists()) {
                        try {
                            Bitmap bitmap = BitmapFactory.decodeFile(imagePath);
                            if (bitmap != null) {
                                background = new BitmapDrawable(activity.getResources(), bitmap);
                            }
                        } catch (Exception e) {
                            AppLogger.error("ThemeManager", "无法加载背景图片: " + e.getMessage());
                        }
                    }
                }
                // 如果图片加载失败，回退到默认背景
                if (background == null) {
                    try {
                        int resId = activity.getResources().getIdentifier("bg_main", "drawable", activity.getPackageName());
                        if (resId != 0) {
                            background = activity.getResources().getDrawable(resId, activity.getTheme());
                        }
                    } catch (Exception e) {
                        AppLogger.error("ThemeManager", "无法加载默认背景: " + e.getMessage());
                    }
                    if (background == null) {
                        background = new ColorDrawable(0xFFFFFFFF);
                    }
                }
                break;
            default:
                // 默认背景
                try {
                    int resId = activity.getResources().getIdentifier("bg_main", "drawable", activity.getPackageName());
                    if (resId != 0) {
                        background = activity.getResources().getDrawable(resId, activity.getTheme());
                    }
                } catch (Exception e) {
                    AppLogger.error("ThemeManager", "无法加载默认背景: " + e.getMessage());
                }
                if (background == null) {
                    background = new ColorDrawable(0xFFFFFFFF);
                }
                break;
        }

        // 应用背景到窗口
        if (background != null && activity.getWindow() != null) {
            activity.getWindow().setBackgroundDrawable(background);
        }
    }

    /**
     * 处理配置变化（主题切换）
     */
    public void handleConfigurationChanged(Configuration newConfig) {
        // 检查是否是 UI 模式改变（深色/浅色模式）
        int currentNightMode = newConfig.uiMode & Configuration.UI_MODE_NIGHT_MASK;
        
        AppLogger.debug("ThemeManager", "配置改变: nightMode=" + currentNightMode);
        
        // 如果设置为"跟随系统"，立即重建Activity以应用主题
        if (settingsManager.getThemeMode() == 0) {
            AppLogger.debug("ThemeManager", "跟随系统模式，重建Activity");
            
            // 先关闭所有对话框，防止recreate后被恢复
            androidx.fragment.app.FragmentManager fm = activity.getSupportFragmentManager();
            for (androidx.fragment.app.Fragment fragment : fm.getFragments()) {
                if (fragment instanceof androidx.fragment.app.DialogFragment) {
                    ((androidx.fragment.app.DialogFragment) fragment).dismissAllowingStateLoss();
                }
            }
            
            // 延迟一点点，确保对话框关闭
            new android.os.Handler().postDelayed(() -> {
                // 重建Activity以应用新主题
                activity.recreate();
            }, 50);
        }
    }
}

File: app/src/main/java/com/app/ralaunch/core/common/ThemeManager.kt
package com.app.ralaunch.core.common

import android.content.res.Configuration
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.fragment.app.DialogFragment
import com.app.ralaunch.core.common.SettingsAccess
import com.app.ralaunch.shared.core.config.BackgroundType
import com.app.ralaunch.shared.core.config.IThemeManager
import com.app.ralaunch.shared.core.config.ThemeConfig
import com.app.ralaunch.shared.core.config.ThemeMode
import com.app.ralaunch.core.common.util.AppLogger

/**
 * 主题管理器 - Android 实现
 * 负责管理主题应用（主题模式、背景设置、动态颜色等）
 * 
 * 实现 shared 模块的 IThemeManager 接口
 */
class ThemeManager(private val activity: AppCompatActivity) : IThemeManager {

    companion object {
        private const val TAG = "ThemeManager"
    }

    private val settingsManager: SettingsAccess = SettingsAccess
    private val dynamicColorManager: DynamicColorManager = DynamicColorManager.getInstance()

    /**
     * 从设置中应用主题（包括深色/浅色模式和动态颜色）
     */
    fun applyThemeFromSettings() {
        applyNightMode()
        applyDynamicColors()
    }

    /**
     * 应用深色/浅色模式
     */
    private fun applyNightMode() {
        when (settingsManager.themeMode) {
            0 -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
            1 -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            2 -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        }
    }

    /**
     * 应用动态颜色主题
     */
    fun applyDynamicColors() {
        try {
            dynamicColorManager.applyDynamicColors(activity)
            AppLogger.info(TAG, "动态颜色主题已应用")
        } catch (e: Exception) {
            AppLogger.error(TAG, "应用动态颜色失败: ${e.message}", e)
        }
    }

    /**
     * 应用自定义主题颜色
     */
    fun applyCustomThemeColor(color: Int) {
        try {
            settingsManager.themeColor = color
            dynamicColorManager.applyCustomThemeColor(activity, color)
            AppLogger.info(TAG, "自定义主题颜色已应用: ${String.format("#%06X", 0xFFFFFF and color)}")
        } catch (e: Exception) {
            AppLogger.error(TAG, "应用自定义主题颜色失败: ${e.message}", e)
        }
    }

    /**
     * 应用背景设置
     * 
     * 注：背景图片和视频由 Compose 层处理（AppThemeState + BackgroundLayer）
     * 此方法仅设置 window 级别的背景色作为底层
     */
    fun applyBackgroundFromSettings() {
        val type = settingsManager.backgroundType
        AppLogger.info(TAG, "applyBackgroundFromSettings - type: $type")

        when (type) {
            "video" -> applyVideoBackground()
            "image" -> applyImageBackground()
            "color" -> applyColorBackground()
            else -> applyDefaultBackground()
        }
    }

    private fun applyVideoBackground() {
        AppLogger.info(TAG, "背景类型: video，设置透明底层（视频由 Compose 渲染）")
        // 视频背景由 Compose 的 VideoBackground 组件处理
        // 这里设置透明背景，让 Compose 层的视频可见
        activity.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
    }

    private fun applyImageBackground() {
        val imagePath = settingsManager.backgroundImagePath
        AppLogger.info(TAG, "背景类型: image，路径: $imagePath（图片由 Compose 渲染）")
        // 图片背景由 Compose 的 BackgroundLayer 组件处理
        // 这里设置透明背景，让 Compose 层的图片可见
        activity.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
    }

    private fun applyColorBackground() {
        val color = settingsManager.backgroundColor
        activity.window?.setBackgroundDrawable(ColorDrawable(color))
        AppLogger.info(TAG, "纯色背景已应用")
    }

    private fun applyDefaultBackground() {
        val nightMode = activity.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
        val background = if (nightMode == Configuration.UI_MODE_NIGHT_YES) {
            ColorDrawable(0xFF121212.toInt())
        } else {
            ColorDrawable(0xFFF5F5F5.toInt())
        }
        activity.window?.setBackgroundDrawable(background)
        AppLogger.info(TAG, "默认纯色背景已应用")
    }

    /**
     * 检查是否使用视频背景
     */
    val isVideoBackground: Boolean
        get() = settingsManager.backgroundType == "video"

    /**
     * 获取视频背景路径
     */
    val videoBackgroundPath: String?
        get() = settingsManager.backgroundVideoPath

    /**
     * 处理配置变化（主题切换）
     */
    fun handleConfigurationChanged(newConfig: Configuration) {
        if (settingsManager.themeMode != 0) return

        // 先关闭所有对话框
        activity.supportFragmentManager.fragments.forEach { fragment ->
            if (fragment is DialogFragment) {
                fragment.dismissAllowingStateLoss()
            }
        }

        // 延迟重建 Activity
        Handler(Looper.getMainLooper()).postDelayed({
            activity.recreate()
        }, 50)
    }

    // ==================== IThemeManager 接口实现 ====================

    override fun getThemeConfig(): ThemeConfig {
        return ThemeConfig(
            mode = ThemeMode.fromValue(settingsManager.themeMode),
            primaryColor = settingsManager.themeColor,
            backgroundType = BackgroundType.fromValue(settingsManager.backgroundType ?: "default"),
            backgroundColor = settingsManager.backgroundColor,
            backgroundImagePath = settingsManager.backgroundImagePath,
            backgroundVideoPath = settingsManager.backgroundVideoPath,
            backgroundOpacity = settingsManager.backgroundOpacity
        )
    }

    override fun setThemeMode(mode: ThemeMode) {
        settingsManager.themeMode = mode.value
        applyNightMode()
    }

    override fun setPrimaryColor(color: Int) {
        applyCustomThemeColor(color)
    }

    override fun setBackgroundType(type: BackgroundType) {
        settingsManager.backgroundType = type.value
        applyBackgroundFromSettings()
    }

    override fun setBackgroundImagePath(path: String?) {
        settingsManager.backgroundImagePath = path ?: ""
    }

    override fun setBackgroundVideoPath(path: String?) {
        settingsManager.backgroundVideoPath = path ?: ""
    }

    override fun setBackgroundOpacity(opacity: Int) {
        settingsManager.backgroundOpacity = opacity
    }

    override fun applyTheme() {
        applyThemeFromSettings()
    }

    override fun applyBackground() {
        applyBackgroundFromSettings()
    }
}
